polkadot/xcm/pallet-xcm/src/benchmarking.rs

```diff
@@ -184,27 +215,40 @@ benchmarks! {
 					&Asset { fun: Fungible(*amount), id: asset.id.clone() },
 					&origin_location,
 					None,
-				).map_err(|error| {
-				  tracing::error!("Fungible asset couldn't be deposited, error: {:?}", error);
-				  BenchmarkError::Override(BenchmarkResult::from_weight(Weight::MAX))
+				)
+				.map_err(|error| {
+					tracing::error!("Fungible asset couldn't be deposited, error: {:?}", error);
+					BenchmarkError::Override(BenchmarkResult::from_weight(Weight::MAX))
+				})?;
+			},
+			NonFungible(_instance) => {
+				<T::XcmExecutor as XcmAssetTransfers>::AssetTransactor::deposit_asset(
+					&asset,
+					&origin_location,
+					None,
+				)
+				.map_err(|error| {
+					tracing::error!("Nonfungible asset couldn't be deposited, error: {:?}", error);
+					BenchmarkError::Override(BenchmarkResult::from_weight(Weight::MAX))
 				})?;
 			},
-			NonFungible(instance) => {
-				<T::XcmExecutor as XcmAssetTransfers>::AssetTransactor::deposit_asset(&asset, &origin_location, None)
-					.map_err(|error| {
-						tracing::error!("Nonfungible asset couldn't be deposited, error: {:?}", error);
-						BenchmarkError::Override(BenchmarkResult::from_weight(Weight::MAX))
-					})?;
-			}
 		};
 
 		let recipient = [0u8; 32];
 		let versioned_dest: VersionedLocation = destination.clone().into();
 		let versioned_beneficiary: VersionedLocation =
 			AccountId32 { network: None, id: recipient.into() }.into();
 		let versioned_assets: VersionedAssets = assets.into();
-	}: _<RuntimeOrigin<T>>(send_origin.into(), Box::new(versioned_dest), Box::new(versioned_beneficiary), Box::new(versioned_assets), 0)
-	verify {
+
+		#[extrinsic_call]
+		_(
+			send_origin,
```

Comment:
It probably still works. However, the `reserve_transfer_assets()` code (see [here](https://github.com/paritytech/polkadot-sdk/blob/adc0178f8702bb619bfebd570edf451f5253d3e0/polkadot/xcm/pallet-xcm/src/lib.rs#L1561)) makes and ensure check on a runtime origin (see [here](https://github.com/paritytech/polkadot-sdk/blob/adc0178f8702bb619bfebd570edf451f5253d3e0/polkadot/xcm/pallet-xcm/src/lib.rs#L238)), so I would still prefer having an explicit `RuntimeOrigin`

Reply:
Hmm, I am not sure if I understand. This is a static type check. There is no "it probably works" as far as I understand it - either the code compiles, or it doesn't. It is actually common to use the `RawOrigin` as you can see in other [benchmarks](https://github.com/paritytech/polkadot-sdk/blob/adc0178f8702bb619bfebd570edf451f5253d3e0/substrate/frame/remark/src/benchmarking.rs#L46).

Just to be clear, this is the code that has been in benchmark v1, which I consider useless.

```rust
			RuntimeOrigin::<T>::from(send_origin) as RuntimeOrigin<T>,
```

And it does the following, it transforms the `RawOrigin` into a `Result<frame_system::RawOrigin<<T as frame_system::Config>::AccountId>, <T as frame_system::Config>::RuntimeOrigin>`, and then we primitive-cast it into a `RuntimeOrigin` with the `as` cast.

We have to do this cast in other benchmark tests in this file because the [T::ExecuteXcmOrigin::try_successful_origin](T::ExecuteXcmOrigin::try_successful_origin) returns such a result from above that we have to cast into an actual origin.

So I honestly think it is a bad idea to change this, we don't have to test any conversions here. I think that the bench v1 code was an error precisely because someone did not understand what happens under the hood and then introduced an `into` followed by a cast just to make the code compile.

Reply:
The macro is actually adding `.into()` to the origin expression:
https://github.com/paritytech/polkadot-sdk/blob/24dab3cfe01c05e621d8a7e26426a068e3e7343c/substrate/frame/support/procedural/src/benchmark.rs#L951-L961

The final type used to dispatch the call is always: `RuntimeOrigin`. Because it is using the trait `UnfilteredDispatchable` implementation on `Call`: https://paritytech.github.io/polkadot-sdk/master/pallet_xcm/pallet/enum.Call.html#impl-UnfilteredDispatchable-for-Call%3CT%3E

IMO the macro is a bit confusing and should just require the expression to return the `RuntimeOrigin` but better not to change it anymore.
So it should be fine to keep the conversion implicit.

Reply:
Wow, thanks a lot for actually deep diving into what is happening under the hood. This is how I should have been arguing - I learned something.

Yeah, I imagine the macro is like this because it allows more flexibility at the price of being harder to understand.

Thanks a lot @re-gius and @gui1117 for the careful review.